### ISSUE TITLE
make flint+arb a dependency?

### ISSUE THREAD

**Issue by rwst** [symengine/symengine#877]
I think flint and arb should be default with SymEngine. There is no alternative when it comes to speed in the domains float/complex and poly/series numerics. There is no reason to have an own additional source of bugs. MPC cannot serve as default complex numerics because it lacks all the special functions.

**Comment by certik** [comment]
This would require SymEngine to change the license from BSD to GPL, and I don't want that. So it needs to stay optional, i.e. SymEngine must build without it if the user wants. Then SymEngine can stay BSD licensed. I asked both @fredrik-johansson and @wbhart a few times to consider using BSD or at least LGPL, but it's apparently not easy due to the various third party contributors.

**Comment by fredrik-johansson** [comment]
Yeah, we have looked at this. Most of the core authors of FLINT have given permission to change the license to LGPL. However, Richard Howell-Peak who wrote the original version of the polynomial factorisation code said no.

Since polynomial factorisation isn't really used by anything else, I think we could make an LGPL edition of FLINT in which the modules for polynomial factorisation have simply been disabled, provided that no other significant contributor vetoes the license change (then there will be more code to swap out).

Bill just created an issue for this, if you want to help out: https://github.com/wbhart/flint2/issues/244

**Comment by certik** [comment]
@fredrik-johansson thank you! That would change everything. @rwst if Flint and Arb is LGPL, then I don't mind depending on them by default. Then the only other issue is to make sure both Flint and Arb build on all the platforms and compilers that we support.

**Comment by wbhart** [comment]
Do you support Windows? Today we discovered that MPFR fails almost all its
tests on Mingw-w64 under msys2, and we depend on MPFR. No idea why.
Probably something stupid we are doing.

There are some failing tests in Arb on Windows (about 5), but maybe that
has to do with the failing MPFR tests. You also currently need to make some
minor changes to the Arb Makefile on Windows.

We don't currently support MIPS or x32.

But we do support x86_64 and x86. Flint ostensibly supports ARM, but I have
a feeling there's a few issues with one of the ARM modes.

We don't support Sun compilers. Clang should be mostly ok, and so should
Apple's GCC and ICC should be ok for the most part. We do support MSVC for
Flint, but not Arb.

I had to back out some BSD patches because they broke everything else.

Unfortunately, we don't have any people we can commit to porting work at
the moment. It's way down the list of priorities.

Bill.

**Comment by fredrik-johansson** [comment]
>  There are some failing tests in Arb on Windows (about 5), but maybe that has to do with the failing MPFR tests. 

Yes, a minor issue with MPFR could easily explain the few failing Arb tests. There shouldn't be any significant problems left on Windows.

> It's way down the list of priorities.

(In other words, capable volunteers are welcome.)

**Comment by isuruf** [comment]
Here are all the compilers that we test regularly.
https://github.com/symengine/symengine/wiki/Compiler-Support

I can help with the porting and also to set up CI solutions as in SymEngine.

**Comment by certik** [comment]
We support Windows (MSVC and mingw), Linux (gcc, clang, intel) and mac
(gcc, clang).  I think we only test 32 bit on windows with mingw.

I would like to support Arm, if we can find some online service to test
each PR. I think we should have a simple android test app, so that people
can take it from there to develop advanced applications.

**Comment by isuruf** [comment]
32 bit MinGW from mingw.org and 64 bit MinGW-w64.

**Comment by fredrik-johansson** [comment]
> Note: I still want to allow an optional BSD / MIT license only build of SymEngine, that means using a slower but BSD licensed equivalent of GMP, and similarly for other libraries.

IMHO, there's little practical benefit to do so since LGPL libraries are effectively just part of the environment (like the OS), as long as you link to them dynamically -- the "SymEngine build" is completely independent of them. Even Maple and Mathematica depend on LGPL libraries, and ship with them.

From the point of view of BSD/MIT software, I think the LGPL is only really problematic in two senses.

The first problem is when targeting embedded systems where static linking is required for technical reasons. This is not really relevant for SymEngine. (Yes, I know that some people claim that statically linked code is 1.35% faster than dynamically linked code. But there are also people who believe that writing ++i instead of i++ makes for loops more aerodynamic.)

The second problem is that you can't copy source code from the LGPL part of the software and put it in the BSD/MIT part of the software. So this limits your flexibility -- the nice thing about a world that is fully BSD/MIT, or a world that is fully copyleft for that matter -- is that you can mix code as you like. But I see this much more as an argument for choosing BSD/MIT when developing new software than an argument against having BSD/MIT licensed code depend on existing LGPL libraries (if the latter is a problem, you should object to running the code on a Linux machine too, let alone the proprietary Intel CPU).

**Comment by rwst** [comment]
> Here are all the compilers that we test regularly.
I just discovered I can't even switch off MPFR in SymEngine so, at least where it is installed(?) on the systems from that list, it already passes all tests.

**Comment by isuruf** [comment]
@rwst, MPFR is not tested on OS X and Windows. You can't turn off MPFR because you must have another option like Flint or Piranha turned on.

**Comment by duhadler** [comment]
Regarding the use of MPIR 2.7.2, MPFR 3.1.4, FLINT 2.5.2 and ARB 2.8.1 on 32 bit and 64 bit Windows, I would like to share my experience on various Windows installations: Windows XP (32 bit), and Windows Vista, 7, 8.1, 10 (32 and 64 bit).

I am using the following system:
Environment: MSYS2 32 bit .
Compiler: TDM-GCC 4.9.2 (12 December 2014). 
The outcome so far is the following:

MPIR and MPFR compile without errors and warnings, and all tests pass (both 32 bit and 64 bit), on MPFR one test (decimals) is skipped.

FLINT 32 bit and 64 bit compile without errors and warnings; there is one test failure in “../build/nmod_poly/test/t-fread_print”, which is caused by the usual lack of compatibility of Windows with the input/output routines of FLINT. The solution is to skip this test (as has already been done in many other cases).

ARB 8.1 currently does not compile correctly on Windows/MSYS2. The reason is that in “../build/acb_hypgeom/acb_hypgeom_airy.c” the transition from long to slong has been missed in line 191. With this corrected, everything compiles. There are no test failures for ARB 32 bit on Windows, but some test failures on ARB 64 bit, all involving logarithms. I have not been able to fully identify the root cause, but it is important to remember that on Windows, the exponent of an mpfr_t is ALWAYS only 32 bits (both for the 32 and 64 bit versions), whereas on Linux it is 64 bit for the 64 bit versions (and 32 bit for the 32 bit versions). Therefore code which explores the full 64 bit exponent range on Linux is likely to fail on Windows. An easy solution would be to limit the test code on 64 bit Windows to cases which are known to work on 32 bit Windows.

In summary, I believe that MPIR and MPFR are fully functional and both FLINT and ARB are very close to be used on 32/64 bit Windows.

Best regards,

Dietrich

P.S.: I am using the following versions and configurations:

MPIR 2.7.2, configured as follows:
32 bit:
./configure --build=i686-w64-mingw32 --host=i686-w64-mingw32 --enable-fat --enable-cxx --enable-static --disable-shared ABI=32 CFLAGS="-m32 -march=i686 -Wno-attributes" --prefix=/local/Win32 --exec-prefix=/local/Win32
64 bit:
./configure --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --enable-fat --enable-static --disable-shared ABI=64 CFLAGS="-m64 -march=k8 -Wno-attributes" --prefix=/local/Win64 --exec-prefix=/local/Win64


MPFR 3.1.4, configured as follows:
32 bit:
./configure --build=i686-w64-mingw32 --host=i686-w64-mingw32 --enable-static --disable-shared CFLAGS="-m32 -march=i686 -Wno-attributes" --with-gmp-include=/local/Win32/include --with-gmp-lib=/local/Win32/lib --prefix=/local/Win32 --exec-prefix=/local/Win32
64 bit:
./configure --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --enable-static --disable-shared CFLAGS="-m64 -march=k8 -Wno-attributes" --with-gmp-include=/local/Win64/include --with-gmp-lib=/local/Win64/lib --prefix=/local/Win64 --exec-prefix=/local/Win64


FLINT 2.5.2, configured as follows:
32 bit:
./configure --build=x86-MINGW32 --enable-static --disable-shared ABI=32 CFLAGS="-m32 -march=i686 -Wno-attributes" --with-gmp=/local/Win32 --with-mpfr=/local/Win32 --prefix=/local/Win32
64 bit:
./configure --build=x86_64-MINGW64 --enable-static --disable-shared ABI=64 CFLAGS="-m64 -march=k8 -Wno-attributes" --with-gmp=/local/Win64 --with-mpfr=/local/Win64 --prefix=/local/Win64


ARB 2.8.1, configured as follows:
32 bit:
./configure --build=x86-MINGW32 --enable-static --disable-shared ABI=32 CFLAGS="-m32 -march=i686 -Wno-attributes" --with-gmp=/local/Win32 --with-mpfr=/local/Win32 --with-flint=/local/Win32 --prefix=/local/Win32
64 bit:
./configure --build=x86_64-MINGW64 --enable-static --disable-shared ABI=64 CFLAGS="-m64 -march=k8 -Wno-attributes" --with-gmp=/local/Win64 --with-mpfr=/local/Win64 --with-flint=/local/Win64 --prefix=/local/Win64

**Comment by fredrik-johansson** [comment]
I wasn't aware that MPFR uses 32 bit exponents on Windows 64. That could certainly explain most (if not all) of the test failures.

I just fixed, hopefully, the one line in arb_log where this is an issue. It needs testing (I don't have a Windows machine to do so myself).

By the way, the MPFR exponent limitation also means that you can't use precision higher than 2^30 bits or so on Windows 64, but that's not urgent to fix.

**Comment by duhadler** [comment]
I have checked it, and indeed, with this change in log.c, now all tests pass on Windows for the 32 bit and 64 bit version of ARB 2.8.1.

There is still a test (arb/test/t-exp_arf_bb.c) in the current arb-master which fails on Windows in 64 bit. This might be a similar issue, looking at line 76 and 78. Everything else is fine.

**Comment by isuruf** [comment]
I started on Travis and Appveyor support for the libraries. https://github.com/wbhart/mpir/pull/171
MPIR/GMP is already okay since we test with them on Travis/Appveyor already.
MPFR is next, we only check MPFR on Linux in SymEngine.

**Comment by fredrik-johansson** [comment]
This would be great to have for flint and arb. I tried to set up travis builds for arb some time ago, but I couldn't figure out how to cache the dependencies (mpir+mpfr+flint). It takes too long to install them from scratch each time.

By default, the flint and arb tests probably take too long to run. When testing flint, flint_test_multiplier() can be changed to return 1 instead of 10. This results in less comprehensive testing, but should be good enough for continuous integration. There's no such multiplier in arb, but it would be easy to add.

**Comment by certik** [comment]
@rwst for now I would like to keep Flint + Arb optional. On your machine, you can always enable it, that's not a big deal.

Rather, what we should do is provide binaries for the common platforms, and there make sure we enable all the libraries that we want to support. I opened a new issue for this: #944

**Comment by rwst** [comment]
Since Flint and Arb have changed to LGPL anything is possible now. The issue stays open accordingly but I'll change the start comment to reflect new realities.

**Comment by isuruf** [comment]
Arb can now be built with MSVC and all tests pass. This is for a static lib and not for a shared lib. There's still some more work to do to support a shared lib.
